zcash_client_backend/src/tor/http.rs

```diff
@@ -0,0 +1,207 @@
+
+        let port = match url.port_u16() {
+            Some(port) => port,
+            None if is_https => 443,
+            None => 80,
+        };
+
+        // Connect to the server.
+        debug!("Connecting through Tor to {}:{}", host, port);
+        let stream = self.inner.connect((host.as_str(), port)).await?;
+
+        if is_https {
+            // On apple-darwin targets there's an issue with the native TLS implementation
+            // when used over Tor circuits. We use Rustls instead.
+            //
+            // https://gitlab.torproject.org/tpo/core/arti/-/issues/715
+            let mut root_store = RootCertStore::empty();
+            root_store.add_trust_anchors(webpki_roots::TLS_SERVER_ROOTS.iter().map(|root| {
+                OwnedTrustAnchor::from_subject_spki_name_constraints(
+                    root.subject,
+                    root.spki,
+                    root.name_constraints,
+                )
+            }));
+            let config = ClientConfig::builder()
+                .with_safe_defaults()
+                .with_root_certificates(root_store)
+                .with_no_client_auth();
+            let connector = TlsConnector::from(Arc::new(config));
+            let dnsname = ServerName::try_from(host.as_str()).expect("Already checked");
+            let stream = connector
+                .connect(dnsname, stream)
+                .await
+                .map_err(HttpError::Tls)?;
+            make_http_request(stream, url, h, f).await
+        } else {
+            make_http_request(stream, url, h, f).await
+        }
+    }
+
+    async fn get_json<T: DeserializeOwned>(&self, url: Uri) -> Result<Response<T>, Error> {
+        self.get(
+            url,
+            |builder| builder.header(hyper::header::ACCEPT, "application/json"),
+            |body| async {
+                Ok(serde_json::from_reader(
+                    body.collect()
+                        .await
+                        .map_err(HttpError::from)?
+                        .aggregate()
+                        .reader(),
+                )
+                .map_err(HttpError::from)?)
+            },
+        )
+        .await
+    }
+}
+
+async fn make_http_request<T, F: Future<Output = Result<T, Error>>>(
+    stream: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    url: Uri,
+    h: impl FnOnce(Builder) -> Builder,
+    f: impl FnOnce(Incoming) -> F,
+) -> Result<Response<T>, Error> {
+    debug!("Making request");
+    let (mut sender, connection) = conn::http1::handshake(TokioIo::new(stream))
+        .await
+        .map_err(HttpError::from)?;
+
+    // Spawn a task to poll the connection and drive the HTTP state.
+    PreferredRuntime::current()?
+        .spawn(async move {
+            if let Err(e) = connection.await {
+                error!("Connection failed: {}", e);
+            }
+        })
+        .map_err(HttpError::from)?;
+
+    let req = h(Request::builder()
```

Comment:
I don't understand the reason for using a callback here. Also, the `.body` call explicitly passing the empty bytes value seems potentially restrictive. 

Is it not possible to just take a `Request` value and the validate it before sending?

Reply:
The callback was an easy way for `get_json` to be self-contained with only JSON-relevant details.

`.body` passing empty bytes is indeed restrictive, but it is sufficient for what we need, and the methods here are all private so we can change these details at any time. I don't know `hyper` well enough to know what the best API is for supporting generic bodies (this changed between `hyper 0.*` and `hyper 1.*`, so my prior experience doesn't help), and I didnd't have time to figure it out.